source/FormationPositioner.cpp

```diff
@@ -0,0 +1,275 @@
+		// the next generate round.
+		shipPositions[ship] = make_pair(relPos, tickTock);
+
+		// Add the ship to the ring.
+		shipsInFormation.push_back(ship->shared_from_this());
+
+		// Trigger immediate re-generation of the formation positions (to
+		// ensure that this new ship also gets a valid position).
+		positionsTimer = 0;
+	}
+
+	return formationLead->Position() + direction.Rotate(relPos);
+}
+
+
+
+// Re-generate the list of (relative) positions for the ships in the formation.
+void FormationPositioner::CalculatePositions()
+{
+	// Set scaling based on results from previous run.
+	double diameterToPx = maxDiameter;
+	double widthToPx = maxWidth;
+	double heightToPx = maxHeight;
+	maxDiameter = 0;
+	maxWidth = 0;
+	maxHeight = 0;
+
+	// Run the position iterator for the ships in the formation.
+	auto itPos = pattern->begin(diameterToPx, widthToPx, heightToPx,
+		centerBodyRadius, shipsInFormation.size());
+
+	// Run the iterator.
+	size_t shipIndex = 0;
+	while(shipIndex < shipsInFormation.size())
+	{
+		// If the ship is no longer valid or not or no longer part of this
+		// formation, then we need to remove it.
+		auto ship = (shipsInFormation[shipIndex]).lock();
+		bool removeShip = !ship || !IsActiveInFormation(ship.get());
+
+		// Lookup the ship in the positions map.
+		auto itCoor = shipPositions.end();
+		if(ship)
+			itCoor = shipPositions.find(&(*ship));
+
+		// If the ship is not in the overall table or if it was not
+		// active since the last iteration, then we also remove it.
+		removeShip = removeShip || itCoor == shipPositions.end() ||
+				itCoor->second.second != tickTock;
+
+		// Perform removes if we need to.
+		if(removeShip)
+		{
+			// Remove the ship from the ring.
+			Remove(shipIndex);
+			// And remove the ship from the shipsPositions map.
+			if(itCoor != shipPositions.end())
+				shipPositions.erase(itCoor);
+		}
+		else
+		{
+			// Set scaling for next round based on the sizes of the
+			// participating ships.
+			Tally(*ship);
+
+			// Calculate the new coordinate for the current ship.
+			Point &shipRelPos = itCoor->second.first;
+			shipRelPos = *itPos;
+			if(flippedY)
+				shipRelPos.Set(-shipRelPos.X(), shipRelPos.Y());
+			if(flippedX)
+				shipRelPos.Set(shipRelPos.X(), -shipRelPos.Y());
+			++itPos;
+			++shipIndex;
+		}
+	}
+
+	// Switch marker to detect stale/missing ships in the next iteration.
+	tickTock = !tickTock;
+}
+
+
+
+void FormationPositioner::Tally(const Body &body)
+{
+	maxDiameter = max(maxDiameter, body.Radius() * 2.);
+	maxHeight = max(maxHeight, body.Height());
+	maxWidth = max(maxWidth, body.Width());
+}
+
+
+
+void FormationPositioner::CalculateDirection()
+{
+	// Calculate new direction, if the formationLead is moving, then we use the movement vector.
+	// Otherwise we use the facing vector.
+	Point velocity = formationLead->Velocity();
+	auto desiredDir = velocity.Length() > 0.1 ? Angle(velocity) : formationLead->Facing();
```

Comment:
I'd just put Angle here
```suggestion
	Angle desiredDir = velocity.Length() > 0.1 ? Angle(velocity) : formationLead->Facing();
```